core/templates/hashfuncs.h

```diff
@@ -437,4 +437,61 @@ const uint32_t hash_table_size_primes[HASH_TABLE_SIZE_MAX] = {
 	1610612741,
 };
 
+// Computed with elem_i = UINT64_C (0 x FFFFFFFF FFFFFFFF ) / d_i + 1, where d_i is the i-th element of the above array.
+const uint64_t hash_table_size_primes_inv[HASH_TABLE_SIZE_MAX] = {
+	3689348814741910324,
+	1418980313362273202,
+	802032351030850071,
+	392483916461905354,
+	190172619316593316,
+	95578984837873325,
+	47420935922132524,
+	23987963684927896,
+	11955116055547344,
+	5991147799191151,
+	2998982941588287,
+	1501077717772769,
+	750081082979285,
+	375261795343686,
+	187625172388393,
+	93822606204624,
+	46909513691883,
+	23456218233098,
+	11728086747027,
+	5864041509391,
+	2932024948977,
+	1466014921160,
+	733007198436,
+	366503839517,
+	183251896093,
+	91625960335,
+	45812983922,
+	22906489714,
+	11453246088
+};
+
+/**
+ * Fastmod computes ( n mod d ) given the precomputed c much faster than n % d.
+ * The implementation of fastmod is based on the following paper by Daniel Lemire et al.
+ * Faster Remainder by Direct Computation: Applications to Compilers and Software Libraries
+ * https://arxiv.org/abs/1902.01961
+ */
+static _FORCE_INLINE_ uint32_t fastmod(const uint32_t n, const uint64_t c, const uint32_t d) {
+#if defined(_MSC_VER)
+	// Returns the upper 64 bits of the product of two 64-bit unsigned integers.
+	// This intrinsic function is required since MSVC does not support unsigned 128-bit integers.
+	return __umulh(c * n, d);
```

Comment:
Does this compile on `x86_32` and ARM?
https://docs.microsoft.com/en-us/cpp/intrinsics/umulh?view=msvc-170 says it's `x64` only.

Reply:
According to https://stackoverflow.com/a/50958815 it might be defined for `ARM64` too (but we should still have a fallback for x86 and ARMv7 - the latter is used for UWP).

Reply:
It is defined for all 64bit platforms. (ARM64: https://docs.microsoft.com/en-us/cpp/intrinsics/arm64-intrinsics?view=msvc-170). However, you're right, it wouldn't compile for x86_32 and ARM32, so I added a fallback for those platforms.